lib/core/utils/get-selector.js

```diff
@@ -1,104 +1,218 @@
-/**
- * Gets the index of element siblings that have the same nodeName
- * Intended for use with the CSS psuedo-class `:nth-of-type()` and xpath node index
- * @param  {HTMLElement} element The element to test
- * @return {Number}         The number of preceeding siblings with the same nodeName
- * @private
- */
-function nthOfType(element) {
-	'use strict';
-
-	var index = 1,
-		type = element.nodeName.toUpperCase();
-
-	/*jshint boss:true */
-	element = element.previousElementSibling;
-	while (element) {
-		if (element.nodeName.toUpperCase() === type) {
-			index++;
-		}
-		element = element.previousElementSibling;
-	}
-
-	return index;
+const escapeSelector = axe.utils.escapeSelector;
+
+function isUncommonClassName (className) {
+  return ![
+    'focus', 'hover',
+    'hidden', 'visible',
+    'dirty', 'touched', 'valid', 'disable',
+    'disable', 'enable', 'active', 'col-'
+  ].find(str => className.includes(str));
 }
 
-/**
- * Checks if an element has siblings with the same selector
- * @param  {HTMLElement} node     The element to test
- * @param  {String} selector The CSS selector to test
- * @return {Boolean}          Whether any of element's siblings matches selector
- * @private
- */
-function siblingsHaveSameSelector(node, selector) {
-	'use strict';
-
-	var index, sibling,
-		siblings = node.parentNode.children;
+function getDistinctClassList (elm) {
+  if (!elm.classList || elm.classList.length === 0) {
+    return [];
+  }
+
+  const siblings = elm.parentNode && Array.from(elm.parentNode.children || '') || [];
+  return siblings.reduce((classList, childElm) => {
+    if (elm === childElm) {
+      return classList;
+    } else {
+      return classList.filter(classItem => {
+        return !childElm.classList.contains(classItem);
+      });
+    }
+  }, Array.from(elm.classList).filter(isUncommonClassName));
+}
 
-	if (!siblings) {
-		return false;
-	}
+const commonNodes = [
+  'div', 'span', 'p', 
+  'b', 'i', 'u', 'strong', 'em',
+  'h2', 'h3'
+];
+
+function getNthChildString (elm, selector) {
+	const siblings = elm.parentNode && Array.from(elm.parentNode.children || '') || [];
+  const hasMatchingSiblings = siblings.find(sibling => (
+  	sibling !== elm &&
+    axe.utils.matchesSelector(sibling, selector)
+  ));
+  if (hasMatchingSiblings) {
+    const nthChild = 1 + siblings.indexOf(elm);
+    return ':nth-child(' + nthChild + ')';
+  } else {
+    return '';
+  }
+}
 
-	var length = siblings.length;
+const createSelector = {
+  // Get ID properties
+  getElmId (elm) {
+  	if (!elm.id) {
+  		return;
+  	}
+  	const id = '#' + escapeSelector(elm.id || '');
+    if (
+    	// Don't include youtube's uid values, they change  on reload
```

Comment:
Yeah, it kinda is. I put it in because YouTube videos are very common, and our video rules will always put them incomplete. If we don't do this, reviewing video's becomes kind of pointless because we can't identify that something is the same video. I'm not in favor of adding as many of these cases as we can find, but this one seemed worth it to me.